src/components/application_manager/src/resumption/resumption_data_processor.cc

```diff
@@ -215,67 +202,100 @@ void ResumptionDataProcessor::ProcessResponseFromHMI(
+    const uint32_t app_id,
+    const hmi_apis::FunctionID::eType function_id,
+    const int32_t corr_id) {
+  sync_primitives::AutoWriteLock lock(resumption_status_lock_);
+  std::vector<ResumptionRequest>& list_of_sent_requests =
+      resumption_status_[app_id].list_of_sent_requests;
 
-  resumption_status_lock_.AcquireForWriting();
   if (resumption_status_.find(app_id) == resumption_status_.end()) {
     LOG4CXX_ERROR(logger_,
                   "No resumption status info found for app_id: " << app_id);
-    resumption_status_lock_.Release();
-    return;
+    return utils::Optional<ResumptionRequest>::OptionalEmpty::EMPTY;
   }
 
-  ApplicationResumptionStatus& status = resumption_status_[app_id];
-  std::vector<ResumptionRequest>& list_of_sent_requests =
-      status.list_of_sent_requests;
-
-  auto request_ptr =
+  auto request_iter =
       std::find_if(list_of_sent_requests.begin(),
                    list_of_sent_requests.end(),
                    [function_id, corr_id](const ResumptionRequest& request) {
                      return request.request_ids.correlation_id == corr_id &&
                             request.request_ids.function_id == function_id;
                    });
 
-  if (list_of_sent_requests.end() == request_ptr) {
+  if (list_of_sent_requests.end() == request_iter) {
     LOG4CXX_ERROR(logger_, "Request not found");
-    resumption_status_lock_.Release();
+    return utils::Optional<ResumptionRequest>::OptionalEmpty::EMPTY;
+  }
+  return utils::Optional<ResumptionRequest>(*request_iter);
+}
+
+utils::Optional<ResumeCtrl::ResumptionCallBack>
+ResumptionDataProcessor::GetResumptionCallback(const uint32_t app_id) {
+  sync_primitives::AutoReadLock lock(register_callbacks_lock_);
+  auto it = register_callbacks_.find(app_id);
+  if (it == register_callbacks_.end()) {
+    LOG4CXX_WARN(logger_, "Callback for app_id: " << app_id << " not found");
+    return utils::Optional<
+        ResumeCtrl::ResumptionCallBack>::OptionalEmpty::EMPTY;
+  }
+  return utils::Optional<ResumeCtrl::ResumptionCallBack>(it->second);
+}
+
+void ResumptionDataProcessor::ProcessResponseFromHMI(
+    const smart_objects::SmartObject& response,
+    const hmi_apis::FunctionID::eType function_id,
+    const int32_t corr_id) {
+  auto found_app_id = GetAppIdByRequestId(function_id, corr_id);
+  if (!found_app_id) {
     return;
   }
+  const uint32_t app_id = *found_app_id;
+  LOG4CXX_DEBUG(logger_, "app_id is: " << app_id);
+
+  LOG4CXX_DEBUG(logger_,
+                "Now processing event with function id: "
+                    << function_id << " correlation id: " << corr_id);
+
+  auto found_request = GetRequest(app_id, function_id, corr_id);
+  if (!found_request) {
+    return;
+  }
+  auto request = *found_request;
+
+  resumption_status_lock_.AcquireForWriting();
+  ApplicationResumptionStatus& status = resumption_status_[app_id];
+  auto& list_of_sent_requests =
```

Comment:
move closer to furst usage

Reply:
Done in 90d5b8f7a4d8ee35f2581eff4cb63a2126a04f08